opensearch/src/main/java/org/opensearch/sql/opensearch/data/type/OpenSearchDataType.java

```diff
@@ -6,82 +6,162 @@
 
 package org.opensearch.sql.opensearch.data.type;
 
-import static org.opensearch.sql.data.type.ExprCoreType.STRING;
-import static org.opensearch.sql.data.type.ExprCoreType.UNKNOWN;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import org.opensearch.sql.data.type.ExprCoreType;
+import org.opensearch.sql.data.type.ExprType;
 
-import com.google.common.collect.ImmutableMap;
 import java.util.Arrays;
-import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import lombok.RequiredArgsConstructor;
-import org.opensearch.sql.data.type.ExprType;
+import java.util.stream.Collectors;
 
 /**
- * The extension of ExprType in Elasticsearch.
+ * The extension of ExprType in OpenSearch.
  */
-@RequiredArgsConstructor
-public enum OpenSearchDataType implements ExprType {
-  /**
-   * OpenSearch Text. Rather than cast text to other types (STRING), leave it alone to prevent
-   * cast_to_string(OPENSEARCH_TEXT).
-   * Ref: https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html
-   */
-  OPENSEARCH_TEXT(Collections.singletonList(STRING), "string") {
-    @Override
-    public boolean shouldCast(ExprType other) {
-      return false;
-    }
-  },
+@EqualsAndHashCode
+public class OpenSearchDataType implements ExprType {
```

Comment:
Should we add different class for each OpenSearch type? Because I think later we need to use them in function signature. For example, OpenSearch storage engine can register function `MATCH(OpenSearchText, ...)`

Reply:
🤔 
Should be a separate class for each OpenSearch type or only for ones which don't have `ExprCoreType` 1:1 mapping?

Reply:
yeah, I think for those special OpenSearch types, we may want to have separate class for each?

Reply:
I'll try this, thank you for the idea